source/common/http/conn_manager_impl.h

```diff
@@ -209,6 +209,25 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void encodeData(Buffer::Instance& data, bool end_stream) override;
     void encodeTrailers(ResponseTrailerMap& trailers) override;
     void encodeMetadata(MetadataMapVector& metadata) override;
+    void setRequestTrailers(Http::RequestTrailerMapPtr&& request_trailers) override {
+      ASSERT(!request_trailers_);
+      request_trailers_ = std::move(request_trailers);
+    }
+    void setContinueHeaders(Http::ResponseHeaderMapPtr&& continue_headers) override {
+      continue_headers_ = std::move(continue_headers);
+    }
+    void setResponseHeaders(Http::ResponseHeaderMapPtr&& response_headers) override {
+      response_headers_ = std::move(response_headers);
+    }
+    void setResponseTrailers(Http::ResponseTrailerMapPtr&& response_trailers) override {
+      response_trailers_ = std::move(response_trailers);
+    }
+    Http::RequestHeaderMap* requestHeaders() override { return request_headers_.get(); }
+    Http::RequestTrailerMap* requestTrailers() override { return request_trailers_.get(); }
+    Http::ResponseHeaderMap* continueHeaders() override { return continue_headers_.get(); }
+    Http::ResponseHeaderMap* responseHeaders() override { return response_headers_.get(); }
+    Http::ResponseTrailerMap* responseTrailers() override { return response_trailers_.get(); }
```

Comment:
nit: prefer optional ref here vs. raw pointer.

Also, FWIW, I really regret passing headers, trailers, etc. in the decode/encode callbacks. I think it would be much better to always force filters to get them through these functions, but that would be a big breaking change for every filter. Might be worth a TODO/comment though?

Reply:
Yeah that seems like a good idea, I'll add a TODO for that